src/lib/operators/table_scan.hpp

```diff
@@ -27,7 +37,100 @@ class TableScan : public AbstractOperator {
   const AllTypeVariant& search_value() const;
 
  protected:
+  template <typename T>
+  class TableScanImpl : public BaseTableScanImpl {
+   public:
+    TableScanImpl(std::shared_ptr<const Table> table, ColumnID column_id, const ScanType scan_type,
+                  const AllTypeVariant search_value);
+
+    std::shared_ptr<const Table> on_execute() override {
+      const auto position_list = std::make_shared<PosList>();
+      auto referenced_table = _table;
+
+      for (auto chunk_id = ChunkID(0); chunk_id < _table->chunk_count(); ++chunk_id) {
+        const auto& chunk = _table->get_chunk(chunk_id);
+        const auto& column = chunk.get_column(_column_id);
+
+        if (auto vc = std::dynamic_pointer_cast<ValueColumn<T>>(column)) {
+          _scanColumn(position_list, vc, chunk_id);
+        } else if (auto dc = std::dynamic_pointer_cast<DictionaryColumn<T>>(column)) {
+          _scanColumn(position_list, dc, chunk_id);
+        } else if (auto rc = std::dynamic_pointer_cast<ReferenceColumn>(column)) {
+          _scanColumn(position_list, rc, referenced_table, chunk_id);
+        } else {
+          throw std::logic_error("Unkown column type.");
+        }
+      }
+
+      const auto result_table = std::make_shared<opossum::Table>();
+
+      Chunk chunk;
+
+      for (auto column_id = ColumnID(0); column_id < _table->col_count(); ++column_id) {
+        result_table->add_column(_table->column_name(column_id), _table->column_type(column_id));
+
+        const auto column = std::make_shared<ReferenceColumn>(referenced_table, column_id, position_list);
+        chunk.add_column(column);
+      }
+
+      result_table->emplace_chunk(std::move(chunk));
+
+      return result_table;
+    }
+
+   protected:
+    template <typename U>
+    bool _evaluate_scan(U value, U compare_value) {
+      switch (_scan_type) {
+        case ScanType::OpEquals:
+          return value == compare_value;
+        case ScanType::OpNotEquals:
+          return value != compare_value;
+        case ScanType::OpGreaterThan:
+          return value > compare_value;
+        case ScanType::OpLessThanEquals:
+          return value <= compare_value;
+        case ScanType::OpGreaterThanEquals:
+          return value >= compare_value;
+        case ScanType::OpLessThan:
+          return value < compare_value;
+        default:
+          throw std::logic_error("Unknown scan type.");
+      }
+    }
+
+    template <typename M, typename N>
```

Comment:
1. `M` and `N` (and `U` above) give no information about what type is expected here. (We know that using `template <typename T>` for one type is common practice and works for one type, but since you are defining 3 template types, this becomes confusing)
2. Instead defining `M`, you could directly expect an accessor argument of type `std::function<N(size_t)>`